apply_kernel.json

```diff
@@ -1,7 +1,7 @@
 {
     "id": "apply_kernel",
     "summary": "Apply a spatial convolution with a kernel",
-    "description": "Applies a 2D convolution (i.e. a focal operation with a weighted kernel) on the horizontal spatial dimensions (axes `x` and `y`) of the data cube.\n\nEach value in the kernel is multiplied with the corresponding pixel value and all products are summed up afterwards. The sum is then multiplied with the factor.",
+    "description": "Applies a 2D convolution (i.e. a focal operation with a weighted kernel) on the horizontal spatial dimensions (axes `x` and `y`) of the data cube.\n\nEach value in the kernel is multiplied with the corresponding pixel value and all products are summed up afterwards. The sum is then multiplied with the factor. The process can't handle no-data values (`null`). The data cube needs to be pre-processed with ``mask()`` or other processes to replace such values as otherwise a `NoDataException` is thrown.",
```

Comment:
Maybe put the part of no-data handling in its own paragraph?

Also: the pre-processing to replace the nans is not that trivial I think (you need at least 3 additional processes: `apply` with `isnan` as subprocess and `mask`). And if you specify that `NoDataException` can thrown, any user will probably want to add this preprocessing to be sure. Wouldn't it be easier (and more efficient on backend side) to specify that `apply_kernel` will automatically replace nan with zero and if the user wants different behaviour, they have to do their own pre-processing explicitly?

Also(2) : what about no-data in the kernel? Here it would be fine to just throw `NoDataException` I think.

Reply:
> Maybe put the part of no-data handling in its own paragraph?

I thought about that, too, and was undecided. So I'll put a line-break in-between.

> Also: the pre-processing to replace the nans is not that trivial I think (you need at least 3 additional processes: `apply` with `isnan` as subprocess and `mask`).

I thought about that, too. ;-) @claxn argued that there's likely a mask anyway so I just put that there. But we can also expand (or remove) it. For now, I just added a larger list of potentially useful processes. Do you have better suggestions?

> And if you specify that `NoDataException` can thrown, any user will probably want to add this preprocessing to be sure. Wouldn't it be easier (and more efficient on backend side) to specify that `apply_kernel` will automatically replace nan with zero and if the user wants different behaviour, they have to do their own pre-processing explicitly?

With @claxn, we thought to add another parameter that allows to set a certain numeric value (e.g. 0 by default) that replaces NA and additionally thought about adding interpolate (see #173). Conclusion was that we don't want to overdo it with this process as then it does quite a lot at a time. Also, we thought it might in many cases not be so useful to just replace with a number and you need a more advanced pre-process step (like interpolate) anyway. What do you think?

> Also(2) : what about no-data in the kernel? Here it would be fine to just throw `NoDataException` I think.

Yes, added that.

Reply:
as a user I would prefer automatic replacing NaN with 0 by `apply_kernel` itself instead of having to think about preparing my data everytime I want to run `apply_kernel`. But maybe other user viewpoints prefer exceptions above automatic/blind replacements.

another solution is introducing a new process like `replace_nan` to replace NaNs with a desired value. 
That would simplify the overhead users have to take care of before running `apply_kernel`

Reply:
I just realized the kernel can't contain a nodata value. Thus, I removed the exception from the code again.

Reply:
And added a replace_nodata parameter now. Can you check it, please?